### ISSUE TITLE
Generally available FIFO

### ISSUE THREAD

**Issue by blyxxyz** [The-Compiler/qutebrowser#1741]
It would be useful to have a FIFO for sending commands, like userscripts get, but that isn't bound to a userscript. It would make it easier to control qutebrowser from a shell, or with scripts that are launched automatically or stay alive longer than the browser itself.

**Comment by The-Compiler** [comment]
I agree. At that point we could probably use that FIFO for all userscripts then (except on Windows, where we use an ordinary file).

**Comment by blyxxyz** [comment]
Using the FIFO for all userscripts would stop them from using the window they were called from.

**Comment by The-Compiler** [comment]
Hmm, that's a good point. I guess commands sent to the generic fifo would always go to the current window?

**Comment by blyxxyz** [comment]
Each FIFO has a `CommandRunner` associated with it, which requires a window id. I'm planning to use `mainwindow.get_window(True)`, so that should behave like the IPC.

**Comment by The-Compiler** [comment]
@blyxxyz so you're working on this?

**Comment by blyxxyz** [comment]
Yes.

**Comment by zpeters** [comment]
Any luck with this?  Sounds like a cool idea

**Comment by The-Compiler** [comment]
@zpeters As you can see, the issue is still open. It's not a priority for me right now, though contributions would be very welcome!

**Comment by zpeters** [comment]
This might be over my head but I'm going to give it a try.  Cheers

**Comment by kskyten** [comment]
@blyxxyz it seems like you have implemented this in your fifo branch. Is it working? If not, what needs to be done?

**Comment by blyxxyz** [comment]
If I remember correctly it sometimes crashed on exit and never worked in tests. I should give it another try.

**Comment by Hackingforlife** [comment]
@The_Compiler, hey there, i would want to take o this issue but i would more clarity on the goal of this issue?

**Comment by jgkamat** [comment]
Hi @Hackingforlife! Here's what I know, let me know if anything is confusing:

When [userscripts](http://qutebrowser.org/doc/userscripts.html) are spawned, qutebrowser creates a special file called a "FIFO". This "FIFO" file allows the userscript to send commands to qutebrowser. The way it works is the program writes to the FIFO as if it was a file, and the contents of the write get sent to qutebrowser, where it is evaluated. For example, if I `echo 'message-info "hello world'" >> $QUTE_FIFO`, I will see 'hello world' appear in qutebrowser. 

Right now, we create this FIFO file only when a userscript is running (I think), this issue is to make that file generally available, so anyone can write to it at any time to run qutebrowser commands.

Luckily, there's a partial implementation of this by @blyxxyz in [this commit](https://github.com/blyxxyz/qutebrowser/commit/c959f74d). However, it's rather stale and probably dosen't work at this point, but it is a good starting point for finding which places may need to be changed. Mainly, I think we need to just initialize the FIFO at start with a predictable name, instead of when a userscript is launched. Let us know if you have any additional questions (we're in `#qutebrowser` on freenode if you want something a bit more real time).
